### PR TITLE
StableDiffusionDepth2ImgPipeline

### PR DESCRIPTION
**Pull request by patil-suraj**
This PR adds `StableDiffusionDepth2ImgPipeline` for image2image using depth masks. Currently it accepts `depth_mask` as input argument, but will soon have depth model inside the pipeline once it's merged in `transformers`.

To run the pipeline temporarily 

```python
!pip install -U git+https://github.com/huggingface/diffusers.git@depth-pipe
!pip install -U git+https://github.com/huggingface/transformers.git

import torch
import requests
from PIL import Image

from diffusers import StableDiffusionDepth2ImgPipeline

pipe = StableDiffusionDepth2ImgPipeline.from_pretrained(
    "stabilityai/stable-diffusion-2-depth",
    torch_dtype=torch.float16,
).to("cuda")


url = "http://images.cocodataset.org/val2017/000000039769.jpg"
init_image = Image.open(requests.get(url, stream=True).raw)
prompt = "two tigers"
n_propmt = "bad, deformed, ugly, bad anotomy"
image = pipe(prompt=prompt, image=init_image, negative_prompt=n_propmt, strength=0.7).images[0]
```

### CODE REVIEW

**Comment by carson-katri** [comment]
Amazing, looking forward to it! I was just [working on a pipeline for this today](https://gist.github.com/carson-katri/f51532b9d5162928d5cacbaee081a799).

**Comment by lawfordp2017** [comment]
Thanks for this,
is the SD2.0 depth2image model converted to diffusers format publicly available?

**Comment by carson-katri** [comment]
@lawfordp2017 i used the conversion script which seemed to work, and uploaded it to [carsonkatri/stable-diffusion-2-depth-diffusers](https://huggingface.co/carsonkatri/stable-diffusion-2-depth-diffusers), until it is available officially.

**Comment by carson-katri** [comment]
Will this pipeline support a passing an pre-existing depth map or is it only for use with MiDaS?

**Comment by patil-suraj** [comment]
The model is trained using dpt-hybrid, so not sure if it will work other depth maps. Have tried using it with other depth maps ? If it works with other depth masks, then we can support passing pre-existing depth maps.

**Comment by patil-suraj** [comment]
That's Awesome, will include an option to pass a external depth map then!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+        if accepts_eta:
+            extra_step_kwargs["eta"] = eta
+
+        # check if the scheduler accepts generator
+        accepts_generator = "generator" in set(inspect.signature(self.scheduler.step).parameters.keys())
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
+    def check_inputs(self, prompt, strength, callback_steps):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:308]: can you use a copied from here as well? Probably from img2img

File: src/diffusers/pipelines/stable_diffusion/__init__.py [hunk 2]
```diff
@@ -39,6 +39,7 @@ class StableDiffusionPipelineOutput(BaseOutput):
 if is_transformers_available() and is_torch_available():
     from .pipeline_cycle_diffusion import CycleDiffusionPipeline
     from .pipeline_stable_diffusion import StableDiffusionPipeline
+    from .pipeline_stable_diffusion_depth2img import StableDiffusionDepth2ImgPipeline
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/__init__.py:42]: Might need the version hack here: `is_transformers_available() and is_torch_available() and is_transformers_version(">=", "4.26.0.dev0")`

**Review by patrickvonplaten — APPROVED** [review]
Can you add docs as well please?

File: docs/source/api/pipelines/stable_diffusion_2.mdx [hunk 3]
```diff
@@ -30,6 +30,8 @@ Note that the architecture is more or less identical to [Stable Diffusion 1](./a
 - *Text-to-Image (768x768 resolution)*: [stabilityai/stable-diffusion-2](https://huggingface.co/stabilityai/stable-diffusion-2) with [`StableDiffusionPipeline`]
 - *Image Inpainting (512x512 resolution)*: [stabilityai/stable-diffusion-2-inpainting](https://huggingface.co/stabilityai/stable-diffusion-2-inpainting) with [`StableDiffusionInpaintPipeline`]
 - *Image Upscaling (x4 resolution resolution)*: [stable-diffusion-x4-upscaler](https://huggingface.co/stabilityai/stable-diffusion-x4-upscaler) [`StableDiffusionUpscalePipeline`]
+- *Text-to-Image (768x768 resolution)*: [stabilityai/stable-diffusion-2](https://huggingface.co/stabilityai/stable-diffusion-2) with [`StableDiffusionPipeline`]
```
Comment (patrickvonplaten) [inline docs/source/api/pipelines/stable_diffusion_2.mdx:33]:
```suggestion
```
```suggestion
- *Text-to-Image (768x768 resolution)*: [stabilityai/stable-diffusion-2](https://huggingface.co/stabilityai/stable-diffusion-2) with [`StableDiffusionPipeline`]
```
```suggestion
- *Text-to-Image (768x768 resolution)*: [stabilityai/stable-diffusion-2](https://huggingface.co/stabilityai/stable-diffusion-2) with [`StableDiffusionPipeline`]
```

Comment (patrickvonplaten) [inline docs/source/api/pipelines/stable_diffusion_2.mdx:33]:
```suggestion
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+
+        depth_map = torch.cat([depth_map] * 2) if do_classifier_free_guidance else depth_map
+        return depth_map
+
+    @torch.no_grad()
+    def __call__(
+        self,
+        prompt: Union[str, List[str]],
+        image: Union[torch.FloatTensor, PIL.Image.Image],
+        depth_map: Optional[torch.FloatTensor] = None,
```
Comment (radames) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:414]: Could the optional depth_map image also be a PIL.Image?

Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:414]: Yes, but for now only accepting tensors, will add support for PIL in next version, need to add a lot of shape checks for that.

Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:414]: For now we can just convert the depth_image into a tensor like this and then pass it to the pipe

```
depth = np.array(depth.convert("L"))
depth = depth.astype(np.float32) / 255.0
depth = depth[None, None]
depth = torch.from_numpy(depth)
```
